src/areastore.cpp

```diff
@@ -0,0 +1,355 @@
+
+#define AST_CONTAINS_PT(a, p) (AST_SMALLER_EQ_AS(a->minedge, p) && \
+	AST_SMALLER_EQ_AS(p, a->maxedge))
+
+#define AST_CONTAINS_AREA(amine, amaxe, b)         \
+	(AST_SMALLER_EQ_AS(amine, b->minedge) \
+	&& AST_SMALLER_EQ_AS(b->maxedge, amaxe))
+
+#define AST_AREAS_OVERLAP(amine, amaxe, b)                \
+	(AST_OVERLAPS_IN_DIMENSION(amine, amaxe, b, X) && \
+	AST_OVERLAPS_IN_DIMENSION(amine, amaxe, b, Y) &&  \
+	AST_OVERLAPS_IN_DIMENSION(amine, amaxe, b, Z))
+
+std::vector<std::string> get_areastore_typenames()
+{
+	std::vector<std::string> res;
+	res.push_back("Vector");
+#if USE_SPATIAL
+	res.push_back("LibSpatial");
+#endif
+	return res;
+}
+
+u16 AreaStore::size() const
+{
+	return areas_map.size();
+}
+
+u32 AreaStore::getFreeId(v3s16 minedge, v3s16 maxedge)
+{
+	//TODO: improve this method:
+	// --> seed random properly at initialisation (eg in constructor)
+	// --> as random will be properly seeded, we won't have to.
+	random.seed(areas_map.size());
+
+	u32 ret;
+	bool free;
+
+	ret = random.next();
+	free = (areas_map.find(ret) == areas_map.end());
+
+	if (free)
+		return ret;
+
+	// we need to seed more stuff
+	random.seed(minedge.X || ((u16)minedge.Y || ((u16)minedge.Z << 16) << 16));
+	random.seed(maxedge.X || ((u16)maxedge.Y || ((u16)maxedge.Z << 16) << 16));
+	
+	int keep_on = 100;
+	while (!free && keep_on--) {
+		ret = random.next();
+		free = (areas_map.find(ret) == areas_map.end());
+	}
+	if (!keep_on) // search failed
+		return 0;
+	return ret;
+}
+
+const Area *AreaStore::getArea(u32 id) const
+{
+	const Area *res = NULL;
+	std::map<u32, Area>::const_iterator itr = areas_map.find(id);
+	if (itr != areas_map.end()) {
+		res = &itr->second;
+	}
+	return res;
+}
+
+bool AreaStore::deserialize(std::istream &is)
+{
+	u8 ver = readU8(is);
+	if (ver != 1)
+		return false;
+	u16 count_areas = readU16(is);
+	for (u16 i = 0; i < count_areas; i++) {
+		// deserialize an area
+		Area a;
+		a.id = readU32(is);
+		a.minedge = readV3S16(is);
+		a.maxedge = readV3S16(is);
+		a.datalen = readU16(is);
+		a.data = new char[a.datalen];
+		is.read((char *) a.data, a.datalen);
+		insertArea(a);
+	}
+	return true;
+}
+
+
+static bool serialize_area(void *ostr, Area *a)
+{
+	std::ostream &os = *((std::ostream *) ostr);
+	writeU32(os, a->id);
+	writeV3S16(os, a->minedge);
+	writeV3S16(os, a->maxedge);
+	writeU16(os, a->datalen);
+	os.write(a->data, a->datalen);
+
+	return false;
+}
+
+
+void AreaStore::serialize(std::ostream &os) const
+{
+	// write initial data
+	writeU8(os, 1); // serialisation version
+	writeU16(os, areas_map.size());
+	forEach(&serialize_area, &os);
+}
+
+
+void AreaStore::invalidateCache()
+{
+	if (cache_enabled) {
+		m_res_cache.invalidate();
+	}
+}
+
+
+void AreaStore::setCacheParams(bool enabled, u8 block_radius, size_t limit)
+{
+	cache_enabled = enabled;
+	m_cacheblock_radius = MYMAX(block_radius, 16);
+	m_res_cache.setLimit(MYMAX(limit, 20));
+	invalidateCache();
+}
+
+inline static s16 other_val(s16 p, u8 r)
```

Comment:
More descriptive method name please (or just don't use an inline function, since it's so short and only used in three places).